### PR TITLE
Proposal: get_traces decorator

### PR DESCRIPTION
**Pull request by alejoe91**
@colehurwitz already added some checks for parsing start and end frame.
I suggest a decorator in the extractor tools that:
- checks `start_frame` and `end_frame`
- checks that all `channel_ids` are correct / filters out wrong ones
- parses to `start_frame` and `end_frame` 

Note that this can be accessed by `spiketoolkit` and could be extended in the future. It would cleanup a lot of the code and fix unwanted bugs (e.g. passing `start_frame` < 0 was not caught before o it needed to be done in each get_traces). 

The `@wraps` decorator makes sure that the docstrings is kept.
If you give me the green light I'll do some cleanup :)

### CODE REVIEW

File: spikeextractors/extraction_tools.py [hunk 1]
```diff
@@ -562,4 +562,82 @@ def load_extractor_from_dict(d):
...
\ No newline at end of file
+    return BaseExtractor.load_extractor_from_dict(d)
+
+
+def check_get_traces_args(func):
+    @wraps(func)
+    def corrected_args(*args, **kwargs):
+        # parse args and kwargs
+        if len(args) == 1:
+            recording = args[0]
```
Comment (samuelgarcia) [inline spikeextractors/extraction_tools.py]: why do you play with arg wheras get_traces is kargs ?

**Comment by samuelgarcia** [comment]
THis is OK for me.
Just one comments.

**Comment by bendichter** [comment]
What checks would it perform? Checking that `start_frame` and `end_frame` are `int`s makes sense, but the downstream error of indexing with non-ints is pretty striaghtforward. 

I don't know about checking that they are positive. I might expect users to intentionally use a negative `start_frame` or `end_frame`, and expects standard `numpy` indexing, i.e. "this many indices from the last index"

**Comment by alejoe91** [comment]
> THis is OK for me.
> Just one comments.

If you call `rec.get_traces([0, 1], 0, 100)`, `channel_ids`, `start_time` and `end_time` are `args`

**Comment by alejoe91** [comment]
> What checks would it perform? Checking that `start_frame` and `end_frame` are `int`s makes sense, but the downstream error of indexing with non-ints is pretty striaghtforward.
> 
> I don't know about checking that they are positive. I might expect users to intentionally use a negative `start_frame` or `end_frame`, and expects standard `numpy` indexing, i.e. "this many indices from the last index"

Hi Ben,
 I see your point. I don't think that all extractors support `np.array` indexing. What I can do is checking that `start_frame` < `end_frame` (that should alwas be true) and if they are negative set them to e.g. `rec.get_num_frames() - end_frame`. Sounds good?

**Comment by bendichter** [comment]
@alejoe91 sounds good to me

**Comment by alejoe91** [comment]
> Hi Ben,
> I see your point. I don't think that all extractors support `np.array` indexing. What I can do is checking that `start_frame` < `end_frame` (that should alwas be true) and if they are negative set them to e.g. `rec.get_num_frames() - end_frame`. Sounds good?

I meant `rec.get_num_frames() + end_frame`

File: spikeextractors/extraction_tools.py (hunk 1 above)
Comment (alejoe91) [reply spikeextractors/extraction_tools.py]: it can be args as well. there is no `*`

Comment (alejoe91) [reply spikeextractors/extraction_tools.py]: It could be args as well as there is no `*`

**Comment by alejoe91** [comment]
Allright fellas merging this

**Pull request merged**